### PR TITLE
feat(gitlab): add users sync

### PR DESCRIPTION
**Pull request by mbliakharskyi**
## Description

Please include a summary of the changes and issues addressed in this PR. Also, explain the motivation and context of each update. List any dependencies or prerequisites needed for implementing this change.

## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: apps/gitlab/src/app/install/route.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  cookies().set('region', region);
+
+  const redirectUrl = new URL(`${env.GITLAB_APP_INSTALL_URL}authorize?`);
+  redirectUrl.searchParams.append('client_id', env.GITLAB_CLIENT_ID);
+  redirectUrl.searchParams.append('redirect_uri', env.GITLAB_REDIRECT_URI);
+  redirectUrl.searchParams.append('response_type', 'code');
+  redirectUrl.searchParams.append('state', organisationId);
+  redirectUrl.searchParams.append('scope', "sudo read_user api read_api"); // Scopes are space-separated.
+  redirectUrl.searchParams.append('code_challenge', "2i0WFA-0AerkjQm4X4oDEhqA17QIAKNjXpagHBXmO_U");
+  redirectUrl.searchParams.append('code_challenge_method', "S256");
```
Comment (elbalexandre) [inline apps/gitlab/src/app/install/route.ts:31]: I don't think PCKE is mandatory, let's not use it :D
```suggestion
```

File: apps/gitlab/src/app/install/route.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  // we store the organisationId in the cookies to be able to retrieve after the SaaS redirection
+  cookies().set('organisation_id', organisationId);
+  cookies().set('region', region);
+
+  const redirectUrl = new URL(`${env.GITLAB_APP_INSTALL_URL}authorize?`);
+  redirectUrl.searchParams.append('client_id', env.GITLAB_CLIENT_ID);
+  redirectUrl.searchParams.append('redirect_uri', env.GITLAB_REDIRECT_URI);
+  redirectUrl.searchParams.append('response_type', 'code');
+  redirectUrl.searchParams.append('state', organisationId);
```
Comment (elbalexandre) [inline apps/gitlab/src/app/install/route.ts:28]: I think state should be a random value, slack integration could serve as example

File: apps/gitlab/src/connectors/auth.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  });
+
+  if (!response.ok) {
+    throw new GitlabError('Could not retrieve token', { response });
+  }
+  const data = (await response.json()) as GetTokenResponseData;
+  return {
+    accessToken: data.access_token,
+    refreshToken: data.refresh_token,
+    expiresIn: data.expires_in/60,
```
Comment (elbalexandre) [inline apps/gitlab/src/connectors/auth.ts:40]: That's for me:
```suggestion
    // TODO: undo /60 when new refresh token flow is implemented 
    expiresIn: data.expires_in/60,
```

File: apps/gitlab/src/connectors/users.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+export type GetUsersParams = {
+  token: string;
+  page: string | null;
+};
+export const getUsers = async ({ token, page }: GetUsersParams) => {
+
+  const url = new URL(`${env.GITLAB_API_BASE_URL}api/v4/users`);
+  url.searchParams.append('pagination', 'keyset');
+  url.searchParams.append('per_page', '2');
```
Comment (elbalexandre) [inline apps/gitlab/src/connectors/users.ts:35]:
```suggestion
  url.searchParams.append('per_page', env.USERS_SYNC_BATCH_SIZE);
```

in env.ts:
```
USERS_SYNC_BATCH_SIZE: z.coerce.number().int().positive().default(2)
```
Could you update the default value with the max value that the gitlab api allows ?

Comment (mbliakharskyi) [reply apps/gitlab/src/connectors/users.ts:35]: Sure, I will do.

File: apps/gitlab/src/app/install/route.ts (hunk 2 above)
Comment (mbliakharskyi) [reply apps/gitlab/src/app/install/route.ts:28]: Got it. Let me check.

File: apps/gitlab/src/app/setup/route.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+export async function GET(request: NextRequest) {
+  const code = request.nextUrl.searchParams.get('code');
+  const organisationId = request.cookies.get('organisation_id')?.value;
+  const region = request.cookies.get('region')?.value;
+
+  console.log("code:", code)
+  console.log("organisationId:", organisationId)
+  if (!organisationId || !code || !region) {
+    redirect(`${env.ELBA_REDIRECT_URL}?error=true`, RedirectType.replace);
+  }
```
Comment (elbalexandre) [inline apps/gitlab/src/app/setup/route.ts:24]:
```suggestion
  const code = request.nextUrl.searchParams.get('code');
  const organisationId = request.cookies.get('organisation_id')?.value;
  const region = request.cookies.get('region')?.value;
  const state = request.searchParams.get('state')
  const cookieState = request.cookies.get('state')?.value;

  console.log("code:", code)
  console.log("organisationId:", organisationId)
  if (!organisationId || !code || !region || state !== cookieState) {
    redirect(`${env.ELBA_REDIRECT_URL}?error=true`, RedirectType.replace);
  }

```

Comment (mbliakharskyi) [reply apps/gitlab/src/app/setup/route.ts:24]: Thanks!

File: apps/gitlab/src/app/install/route.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  cookies().set('organisation_id', organisationId);
+  cookies().set('region', region);
+  cookies().set('state', state);
+  
+  const redirectUrl = new URL(`${env.GITLAB_APP_INSTALL_URL}authorize?`);
+  redirectUrl.searchParams.append('client_id', env.GITLAB_CLIENT_ID);
+  redirectUrl.searchParams.append('redirect_uri', env.GITLAB_REDIRECT_URI);
+  redirectUrl.searchParams.append('response_type', 'code');
+  redirectUrl.searchParams.append('state', state);
+  redirectUrl.searchParams.append('scope', "sudo read_user api read_api"); // Scopes are space-separated.
```
Comment (elbalexandre) [inline apps/gitlab/src/app/install/route.ts:31]:
```suggestion
  redirectUrl.searchParams.append('scope', "read_user api read_api"); // Scopes are space-separated.

```

I am pretty sure we don't need the sudo permissions (it's works without it)

Comment (mbliakharskyi) [reply apps/gitlab/src/app/install/route.ts:31]: okay.

**Review by elbalexandre — APPROVED** [review]

**Pull request merged**